src/systemdhook.c

```diff
@@ -58,6 +68,113 @@ DEFINE_CLEANUP_FUNC(yajl_val, yajl_tree_free)
 
 #define CGROUP_ROOT "/sys/fs/cgroup"
 
+static int makepath(char *dir, mode_t mode)
+{
+    if (!dir) {
+	errno = EINVAL;
+	return -1;
+    }
+
+    if (strlen(dir) == 1 && dir[0] == '/')
+	return 0;
+
+    makepath(dirname(strdupa(dir)), mode);
+
+    return mkdir(dir, mode);
+}
+
+static int bind_mount(const char *src, const char *dest, const char *opt) {
+	if (mount(src, dest, "bind", MS_BIND|MS_REC, NULL) == -1) {
+		pr_perror("Failed to mount %s on %s", src, dest);
+		return -1;
+	}
+	if (mount(src, dest, "bind", MS_REMOUNT|MS_BIND|MS_REC, opt) == -1) {
+		pr_perror("Failed to remount %s", dest);
+		return -1;
+	}
+	return 0;
+}
+
+/* error callback */
+static int parser_errcb(struct libmnt_table *tb __attribute__ ((__unused__)),
+			const char *filename, int line)
+{
+	pr_perror("%s: parse error at line %d", filename, line);
+	return 0;
+}
+
+static struct libmnt_table *parse_tabfile(const char *path)
+{
+	int rc;
+	struct libmnt_table *tb = mnt_new_table();
+
+	if (!tb) {
+		pr_perror("failed to initialize libmount table");
+		return NULL;
+	}
+
+	mnt_table_set_parser_errcb(tb, parser_errcb);
+
+	rc = mnt_table_parse_file(tb, path);
+
+	if (rc) {
+		mnt_free_table(tb);
+		pr_perror("can't read %s", path);
+		return NULL;
+	}
+	return tb;
+}
+
+/* reads filesystems from @tb (libmount) looking for cgroup file systems
+   then bind mounts these file systems over rootfs
+ */
+static int mount_cgroup(struct libmnt_table *tb,
+			struct libmnt_fs *fs,
+			const char *rootfs)
+{
+	struct libmnt_fs *chld = NULL;
+	_cleanup_mnt_iter_ struct libmnt_iter *itr = NULL;
+
+	if (!fs) {
+		/* first call, get root FS */
+		if (mnt_table_get_root_fs(tb, &fs))
+			return -1;
+
+	}
+
+	itr = mnt_new_iter(MNT_ITER_FORWARD);
+	if (!itr)
+		return -1;
+
+	/*
+	 * add all children to the output table
+	 */
+	while(mnt_table_next_child_fs(tb, itr, fs, &chld) == 0) {
+		if (strncmp(mnt_fs_get_target(chld), "/sys/fs/cgroup",14) == 0) {
+			const char *src = mnt_fs_get_target(chld);
+			char dest[PATH_MAX];
+			snprintf(dest, PATH_MAX, "%s%s", rootfs, src);
+
+			if (makepath(dest, 0755) == -1) {
+				if (errno != EEXIST) {
+					pr_perror("Failed to mkdir container cgroup dir");
+					return -1;
+				}
+			}
+			const char *opt="ro";
```

Comment:
I am wondering whether we should use MS_RDONLY flag here instead of the data option here. I think we use the flag in libcontainer.